### PR TITLE
feat(storage): Introduces a split implementation based on split_key to replace the previous move implementation.

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Previously, we implemented the concept of compaction groups for hummock to achieve write/configuration isolation and to create new compaction groups via the table-level move function.

Since the sst key range generated by the move function is not contiguous, this makes it difficult to re-combine the resulting independent compaction groups (currently not supported). Because of this concern, we set the move threshold very high, fearing that the resulting compaciton group will not be able to be merged again, thus creating a large number of compaction groups.

This concern is contrary to our expectation that in scenarios with high write throughput (backfill, etc.), we would like to split the compaction to achieve parallel compaction, increase the overall compaction throughput, and make full use of the compactor node's scale capability.

 Here are the changes involved in this PR for review.
1. based on `table_id` level split function, compatible with the original Split interface.
2. fix key_range/table_id/sst_size after sst split.
3. Use (table_id, vnode = 0, epoch = MAX_REPRESENTABLE) as split key for splitting.

I'll summarise the changes to the file, for your code reivew:

1. src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs introduce helper functions that facilitate splitting, refactor the split interface based on the new implementatio
2. src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs reimplemented split logic based on split_key.
3. src/storage/hummock_sdk/src/compaction_group/mod.rs provides a wrapper for the split_key based sst / level split implementation.

There may be more code to change in this PR, but the logic is more homogeneous, thank you very much for your review!

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 1]
```diff
@@ -357,3 +361,327 @@ impl HummockManager {
...
+                result_vec[1].0
+            }
+        };
+        for (cg_id, table_ids) in result_vec {
+            cg_id_to_table_ids.insert(cg_id, table_ids);
+        }
+
+        // `partition_vnode_count` only works inside a table, to avoid a lot of slicing sst, we only enable it in groups with high throughput and only one table.
+        // The target `table_ids` might be split to an existing group, so we need to try to update its config
+        if table_ids.len() == 1 {
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:662]: Do we need this logic for all CG with single table? If yes, only checking the provided `table_ids` is not enough. Consider the following example:
```
original cg table_ids = [1, 2, 3]
split table_ids = [2]

resulting cg1 table_ids = [1]
resulting cg2 table_ids = [2]
resulting cg3 table_ids = [3]
```

All the resulting CGs need this logic. I wonder whether we should do this inside `split_compaction_group_impl`

Comment (hzxa21) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:662]: Another issue if we do it outside of `split_compaction_group_impl`: the compaction group txn will not be committed within the same transaction of version delta, which may cause in consistency when error happens.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 2]
```diff
@@ -357,3 +361,327 @@ impl HummockManager {
...
+            }
+        };
+        for (cg_id, table_ids) in result_vec {
+            cg_id_to_table_ids.insert(cg_id, table_ids);
+        }
+
+        // split 2
+        // See the example in L603 and the split rule in `split_compaction_group_impl`.
+        let result_vec = self
+            .split_compaction_group_impl(
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:642]: Do we need to make sure split1 and split2 happens atomically in the same txn? If not, what happen if split1 succeeds but split2 fails.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 3]
```diff
@@ -357,3 +361,327 @@ impl HummockManager {
...
+        assert!(result_vec.len() <= 2);
+        let target_compaction_group_id = {
+            if result_vec[0].1.contains(&table_ids[0]) {
+                result_vec[0].0
+            } else {
+                result_vec[1].0
+            }
+        };
+        for (cg_id, table_ids) in result_vec {
+            cg_id_to_table_ids.insert(cg_id, table_ids);
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:657]: Let's add the following assertions:
1. table_ids in different cg are sorted.
2. table_ids in different cg are non-overlapping.
3. the provided table_ids belong to one and only one cg.

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 4]
```diff
@@ -597,12 +626,27 @@ impl HummockVersion {
...
-                    parent_group_id,
-                    *compaction_group_id,
-                    member_table_ids,
-                    group_construct.get_new_sst_start_id(),
-                );
+                if group_construct.version >= CompatibilityVersion::SplitGroupByTableId as _ {
+                    let split_key = if group_construct.split_key.is_some() {
+                        Some(Bytes::from(group_construct.split_key.clone().unwrap()))
+                    } else {
+                        None
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:633]: Is this branch reachable?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 5]
```diff
@@ -154,4 +358,123 @@ pub mod group_split {
...
+
+            let sst = &mut level.table_infos[pos];
+            let sst_split_type = need_to_split(sst, split_key.clone());
+            match sst_split_type {
+                SstSplitType::Left => {
+                    insert_table_infos.extend_from_slice(&level.table_infos[pos + 1..]);
+                    level.table_infos = level.table_infos[0..=pos].to_vec();
+                }
+                SstSplitType::Right => {
+                    insert_table_infos.extend_from_slice(&level.table_infos[pos..]); // the sst at pos has been split to the right
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:410]: If we reach here, pos must be 0? Can we add an assertion for this?

**Review by hzxa21 — COMMENTED** [review]
Thanks for the efforts!

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 6]
```diff
@@ -357,3 +361,327 @@ impl HummockManager {
...
+    /// Split `table_ids` to a dedicated compaction group.
+    /// Returns the compaction group id containing the `table_ids`.
+    pub async fn split_compaction_group(
+        &self,
+        parent_group_id: CompactionGroupId,
+        table_ids: &[StateTableId],
+        partition_vnode_count: u32,
+    ) -> Result<CompactionGroupId> {
+        let (target_compaction_group_id, _) = self
+            .move_state_tables_to_dedicated_compaction_group(
```
[the remark this answers is not in the document]
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:375]: In contrast,  I think the interface `move_state_tables_to_dedicated_compaction_group` should be retained, as the semantics of the interface is still `move`, and I intend to change its semantics when we don't need to move. (After this PR)

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 4 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:633]: https://github.com/risingwavelabs/risingwave/blob/fd8f4b013c03542918444272dbd2fec4bb58f5f1/src/meta/src/hummock/manager/transaction.rs#L153

reachable when construct a new compaction group

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 1 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:662]: It is not required that all single table cgs must be partitioned, we just need to enable it for high throughput cgs, we could even consider deprecating this feature, but for the current PR we choose to keep it. Let me think about how to implement it atomically.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 2 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:642]: Actually, from my point of view it doesn't need to be. 

If split 1 succeeds and split 2 fails (split condition/network), we just wait for the next split to be triggered, if necessary.

File: src/storage/hummock_sdk/src/compaction_group/mod.rs (hunk 5 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs:410]: I don't fully understand what you mean, L409's SstSplitType corresponds to the type of the sst at the pos. 

The value of pos depends on the values of table_infos and split_key in L397, maybe I misunderstood you?

Comment (hzxa21) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs:410]: We use `get_split_pos` to find the last SST with `key_range_left < split_key` if any or return pos 0. If `need_to_split` returns `SstSplitType::Right`, that means `split_key >= key_range_left`. The only possibilities here is pos==0.

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 7]
```diff
@@ -46,11 +46,215 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+        if VirtualNode::MAX == vnode {
+            // Modify `table_id` to `next_table_id` to satisfy the `split_to_right`` rule, so that the `table_id`` originally passed in will be split to left.
+            table_id = table_id.strict_add(1);
+            vnode = VirtualNode::ZERO;
+        }
+
+        FullKey::new(
+            TableId::from(table_id),
+            TableKey(vnode.to_be_bytes().to_vec()),
+            HummockEpoch::MIN,
```
Comment (zwang28) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:88]: Is it HummockEpoch::MAX?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs (hunk 5 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs:413]: Got it, add assertion

**Comment by Li0k** [comment]
PTAL @zwang28 @hzxa21

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 8]
```diff
@@ -837,6 +881,122 @@ impl HummockVersion {
...
+                } else {
+                    warn!(
+                        new_sst_start_id,
+                        "non-zero sst start id for NewCompactionGroup"
+                    );
+                }
+            }
+            return;
+        } else if !self.levels.contains_key(&parent_group_id) {
+            warn!(parent_group_id, "non-existing parent group id to init from");
```
Comment (zwang28) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:909]: When will this happen? Is it enough to WARN?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 9]
```diff
@@ -46,11 +46,217 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+    ) -> SstableInfo {
+        let mut branch_table_info = origin_sst_info.clone();
+        branch_table_info.sst_id = *new_sst_id;
+        *new_sst_id += 1;
+        origin_sst_info.sst_id = *new_sst_id;
+        *new_sst_id += 1;
+
+        let (key_range_l, key_range_r) = {
+            let key_range = &origin_sst_info.key_range;
+            let l = KeyRange {
```
Comment (zwang28) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:153]: In the context of split_sst, the relative ordering of sst.left,sst.right and split_key cannot be assumed.
Maybe assert l.left < l.right and r.left<=r.right, or handle it properly?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs (hunk 7 above)
Comment (hzxa21) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs]: Please update the PR description and relevant docs in the codes as well.

File: src/storage/hummock_sdk/src/compaction_group/mod.rs (hunk 9 above)
Comment (zwang28) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs:153]: e.g. split_key = sst.left = sst.right (inclusive).
We don't want any resulted sst.left > sst.right.

Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/mod.rs:153]: Good catch, refactor `split_sst` and add related UT

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 10]
```diff
@@ -415,19 +437,31 @@ pub mod group_split {
...
+                        insert_table_infos.push(branch_sst);
+                    }
+
+                    let right_start = pos + 1;
+                    let mut left_end = pos;
+
+                    if let Some(origin_sst) = left {
+                        *sst = origin_sst;
+                    } else {
+                        left_end -= 1;
```
Comment (zwang28) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:458]: subtract with overflow?

**Review by zwang28 — APPROVED** [review]
LGTM

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 11]
```diff
@@ -154,4 +375,143 @@ pub mod group_split {
...
+            let sst_split_type = need_to_split(&level.table_infos[pos], split_key.clone());
+            match sst_split_type {
+                SstSplitType::Left => {
+                    insert_table_infos.extend_from_slice(&level.table_infos[pos + 1..]);
+                    level.table_infos = level.table_infos[0..=pos].to_vec();
+                }
+                SstSplitType::Right => {
+                    assert_eq!(0, pos);
+                    insert_table_infos.extend_from_slice(&level.table_infos[pos..]); // the sst at pos has been split to the right
+                    level.table_infos = level.table_infos[0..pos].to_vec();
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:435]: nits: can be `level.table_infos.clear()`

**Review by hzxa21 — APPROVED** [review]
LGTM. Please check whether the PR description is accurate. For example, the split key now uses MAX instead of 0.

**Pull request merged**